codalab/model/bundle_model.py

```diff
@@ -258,11 +258,59 @@ def get_children_uuids(self, uuids):
             result[row.parent_uuid].append(row.child_uuid)
         return result
 
+    def get_n_host_worksheet_uuids(self, bundle_uuids, n):
+        """
+        Get up to n host_worksheet uuids per bundle uuid. n of 0 or less will return an empty dictionary.
+        bundle_uuids: list of bundle uuid's (e.g. ['0x12345', '0x23456'])
+        n: max limit of host_worksheet uuid's to fetch per bundle
+        Return dict of bundle uuid's to a list of host worksheet uuid's {'0x12345': [host_worksheet_uuid, ...], ...}
+        """
+        if n <= 0:
+            return dict()
+        if n == 1:
+            result = self.get_single_host_worksheet_uuid(bundle_uuids)
+            return dict(
+                (uuid, [worksheet]) for uuid, worksheet in result.items() if worksheet is not None
+            )
+
+        with self.engine.begin() as connection:
```

Comment:
@percyliang: I looked into lots of different approaches and If we wanted to try `n` different worksheet uuid's per bundle, I think this is the best solution. Other solutions require `order_by` worksheet_uuids across the bundles.

Reply:
Don't we need to put a limit in the actual query so that if a bundle shows up on a ton of worksheets but we're only getting 5, we're not too wasteful?  Maybe something in here would help? https://www.xaprb.com/blog/2006/12/07/how-to-select-the-firstleastmax-row-per-group-in-sql/

Reply:
I'm worried about a performance hit since I believe this logic is going to be exercised on every call now...can you provide some indication of the change in performance (on dev, where there's a lot of data)?

Reply:
@percyliang: On my local machine, I created ~1000 bundles, where each bundle has ~1000 host worksheets.

- Single worksheet retrieval + checking read permissions took ~0.055 seconds on average.
- Retrieving all but limiting to 5 worksheets + checking read permissions took ~0.078 seconds on average.
- Retrieving all (no limit) + checking read permissions took ~0.4 seconds on average.

I'm leaning towards the limiting to 5 worksheets approach. It will most likely yield correct results to the user and fetching all worksheet uuid's for multiple bundles was just a bit slower than the group by.

Reply:
Cool.  But does the current logic actually retrieve 5 *per* bundle (see my first comment)?  I wonder still if we can just have a single function instead of two (n=1 and n>1)?

Reply:
@percyliang: Unfortunately it doesn't. It still retrieves all. I just limit the read permission checks to 5, making it faster than checking all. 

Next step, I can try `order by` + `limit` to get exactly 5 and measure its performance compared to the current implementations. I speculate it will be slower because of the `order-by`. If it's fast, I can use it to handle both n=1 and n>1 cases.

Reply:
@percyliang: I discovered a new solution that's cleaner and faster than the previous solutions. For 5 worksheets + checks it takes about ~0.041 seconds on average.